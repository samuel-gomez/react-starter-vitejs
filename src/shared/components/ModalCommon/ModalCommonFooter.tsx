import { Button, Modal } from '@axa-fr/react-toolkit-all';
import type { MouseEventHandler, ReactNode } from 'react';

export type TModalCommonFooter = {
  onCancel?: MouseEventHandler<HTMLButtonElement>;
  onSubmit?: MouseEventHandler<HTMLButtonElement>;
  cancelLabel?: ReactNode;
  confirmLabel?: ReactNode;
  hasErrors?: boolean;
  confirmClassModifier?: string;
  cancelClassModifier?: string;
};

const ModalCommonFooter = ({
  onSubmit,
  onCancel,
  cancelLabel = 'Annuler',
  confirmLabel = 'Confirmer',
  hasErrors = false,
  confirmClassModifier = 'success',
  cancelClassModifier = 'reverse',
}: TModalCommonFooter) => (
  <Modal.Footer>
    {onCancel && (
      <Button className="btn af-btn" onClick={onCancel} classModifier={cancelClassModifier}>
        <span className="af-btn__text">{cancelLabel}</span>
      </Button>
    )}
    {onSubmit && (
      <Button disabled={!!hasErrors} className="btn af-btn" classModifier={`hasiconRight ${confirmClassModifier}`} type="button" onClick={onSubmit}>
        <span className="af-btn__text">{confirmLabel}</span>
        <i className="glyphicon glyphicon-arrowthin-right" />
      </Button>
    )}
  </Modal.Footer>
);

export default ModalCommonFooter;
