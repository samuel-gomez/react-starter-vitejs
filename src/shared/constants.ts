import packageJson from '../../package.json';

/** ***********************************************************************
 * FORM CONSTANTS
 ************************************************************************ */
export const MSG_REQUIRED = 'Le champ est obligatoire';

export const MSG_LENGTH_3 = 'Le champ doit faire au moins 3 caractères';
export const MSG_FORMAT = 'Format invalide';
export const MSG_ERROR_FORMAT = 'Format de fichier incorrect';

export const DEFAULT_OPTION_LABEL = '- Sélectionner -';
export const DEFAULT_OPTION = {
  value: '',
  label: DEFAULT_OPTION_LABEL,
};

// patterns
export const patternEmail = /^[\w-+]+@([\w-+]+\.)+[\w-]{2,}$/;
export const patternPhoneNumber = /^[+0][\d ]*$/;
export const patternZipCode = /^\d{5}$/;
export const patternSiret = /^\d{14}$/;
export const patternSiren = /^\d{9}$/;
export const patternSiretSiren = /^\d{9,14}$/;

export const phoneNumberRule = {
  required: true,
  minLength: 4,
  maxLength: 20,
  pattern: patternPhoneNumber,
};

export const emailRule = {
  required: true,
  minLength: 6,
  maxLength: 60,
  pattern: patternEmail,
};

/** ***********************************************************************
 * API STATUS CONSTANTS
 ************************************************************************ */
export const STATUS_HTTP = {
  SUCCESS: 200,
  CREATED: 201,
  BAD_REQUEST: 400,
  UNAUTHORIZE: 401,
  FORBIDDEN: 403,
  NOTFOUND: 404,
  SERVER_ERROR: 500,
};

export const STATUS_API = {
  SUCCESS: STATUS_HTTP.SUCCESS,
  CREATED: STATUS_HTTP.CREATED,
  WARNING: 40,
  ERROR: 50,
};

/** ***********************************************************************
 * API URL
 ************************************************************************ */
export const API_URL = {
  BASE: 'base',
  GITHUB: 'github',
};

export const STATUS_HTTP_MESSAGES = {
  [STATUS_HTTP.SUCCESS]: 'Succès: La requête a réussi',
  [STATUS_HTTP.CREATED]: 'Succès: Création réussie',
  [STATUS_HTTP.BAD_REQUEST]: 'Erreur: La syntaxe de la requête est erronée',
  [STATUS_HTTP.UNAUTHORIZE]: 'Erreur: Une authentification est nécessaire pour accéder à la ressource',
  [STATUS_HTTP.FORBIDDEN]: 'Erreur: Accès non autorisé',
  [STATUS_HTTP.NOTFOUND]: 'Erreur: Elément non trouvé',
  [STATUS_HTTP.SERVER_ERROR]: 'Erreur: Problème technique ! Contacter votre support',
};

/** ***********************************************************************
 * USER CONSTANTS
 ************************************************************************ */

export const ADMIN = 'Admin';
export const READER = 'Reader';
export const WRITER = 'Writer';
export const USER = 'User';
export const ALL = '';

export const PROFILS = [ADMIN, READER, WRITER, USER, ALL];

/** ***********************************************************************
 * COMMONS CONSTANTS
 ************************************************************************ */

export const MODIFIER_CLASS = {
  success: 'success',
  disabled: 'disabled',
};

/** ***********************************************************************
 * VERSION CONSTANT
 ************************************************************************ */

const VERSION_TOOLKIT = 'master';

/** ***********************************************************************
 * DEMO CONSTANTS
 ************************************************************************ */

export const STORYBOOK = 'https://axafrance.github.io/react-toolkit/latest/storybook/?path=/story/';
export const DESIGN_SYSTEM = 'https://axafrance.github.io/design-system/';
export const GITHUB = `https://github.com/AxaGuilDEv/react-toolkit/tree/v${packageJson.dependencies['@axa-fr/react-toolkit-all']}/packages/`;
export const GITHUB_API = `AxaGuilDEv/react-toolkit/${VERSION_TOOLKIT}/packages/`;
