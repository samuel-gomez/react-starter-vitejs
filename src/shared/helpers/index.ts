export { formatDate, isValidDate, setDate } from './formatDate';
export { default as isEmptyOrNull, isEmptyArray, isEmptyObject } from './isEmptyOrNull';
export { default as mergeObj } from './mergeObj';
export { default as setAnomalyEmptyItems } from './setAnomalyEmptyItems';
export { default as setConfirmClassModifier } from './setConfirmClassModifier';
export { default as emptyFunction } from './emptyFunction';
export { default as setRequiredFieldRule, setRequiredMessage } from './setRequiredFieldRule';
