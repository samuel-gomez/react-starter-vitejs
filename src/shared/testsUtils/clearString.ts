const cleanString = (str?: string) => str?.replace(/\s/g, '').replace(/\n/g, '').replace(/\t/g, '').trim();

export default cleanString;
