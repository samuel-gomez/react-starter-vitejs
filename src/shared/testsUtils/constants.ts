const MOCK_API_URL = { base: 'https://your-url-api/' };

export default MOCK_API_URL;
