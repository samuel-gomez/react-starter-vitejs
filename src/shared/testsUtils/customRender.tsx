import { render, renderHook, type RenderOptions } from '@testing-library/react';
import EnvironmentProvider from 'App/EnvironmentProvider';
import FetchProvider from 'App/FetchProvider';
import NotificationProvider from 'App/NotificationProvider';
import QueryProvider from 'App/QueryProvider';
import UserProvider from 'App/UserProvider';
import { type ReactElement, type ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';
import MOCK_API_URL from './constants';

type TMockProvider = {
  [x: string]: Record<string, unknown | number | string> | string | boolean | null | number | Record<string, number | string>[];
};

const MockProviders =
  ({
    role = '',
    isEnabled = true,
    route = '/',
    oidcUser = { member_of: [`CN=${role}`], name: 'Samuel Gomez' },
    accessToken = 'accessToken',
    queryData,
    headers = {
      'Content-type': 'application/json; charset=UTF-8',
    },
  }: TMockProvider) =>
  ({ children }: { children: ReactNode }) => {
    const getAccessTokenFn = vi.fn().mockReturnValue(() => ({ accessToken }));
    const getUserInfosFn = vi.fn().mockReturnValueOnce(() => ({ oidcUser }));
    const useEnvFn = vi.fn().mockReturnValueOnce({
      envState: {
        environment: {
          apiUrl: MOCK_API_URL,
          fetchConfig: {
            headers,
          },
          oidc: {
            isEnabled,
          },
        },
      },
    });

    const queriesOptions = {
      retry: false,
      ...(queryData ? { queryFn: () => queryData } : {}),
    };

    return (
      <EnvironmentProvider useEnvFn={useEnvFn}>
        <UserProvider getUserInfosFn={getUserInfosFn}>
          <FetchProvider getAccessTokenFn={getAccessTokenFn}>
            <QueryProvider queriesOptions={queriesOptions}>
              <NotificationProvider>
                <MemoryRouter initialEntries={[`${route}`]}>{children}</MemoryRouter>
              </NotificationProvider>
            </QueryProvider>
          </FetchProvider>
        </UserProvider>
      </EnvironmentProvider>
    );
  };

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>, testMock: TMockProvider = { role: '' }) =>
  render(ui, { wrapper: MockProviders(testMock), ...options });

type TcustomRenderHook = (testMock?: TMockProvider) => typeof renderHook;

const customRenderHook: TcustomRenderHook = testMock => (hook, options) => renderHook(hook, { wrapper: MockProviders(testMock ?? {}), ...options });

// re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { axe } from 'jest-axe';

// override render method
export { customRenderHook, customRender as render };
