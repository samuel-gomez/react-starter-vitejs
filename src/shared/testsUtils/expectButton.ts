import { screen, within } from '@testing-library/dom';

type TexpectButton = {
  name: string;
  isQueryByRole?: boolean;
  beDisabled?: boolean;
  beInDoc?: boolean;
  parentLabel?: string;
};

const expectButton = ({ isQueryByRole = false, name, beDisabled = true, beInDoc = true, parentLabel = '' }: TexpectButton) => {
  const base = parentLabel ? within(screen.getByLabelText(parentLabel)) : screen;
  const button = isQueryByRole ? base.queryByRole('button', { name: RegExp(name) }) : base.getByRole('button', { name: RegExp(name) });
  if (beInDoc) {
    expect(button).toBeInTheDocument();
    if (beDisabled) {
      expect(button).toBeDisabled();
    } else {
      expect(button).not.toBeDisabled();
    }
  } else {
    expect(button).not.toBeInTheDocument();
  }
};

export default expectButton;
