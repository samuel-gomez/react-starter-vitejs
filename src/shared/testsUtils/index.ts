export * from './customRender';
export { default as emptyFunction } from '../helpers/emptyFunction';
export { default as MOCK_API_URL } from './constants';
export { default as renderWithContainer } from './renderWithContainer';
