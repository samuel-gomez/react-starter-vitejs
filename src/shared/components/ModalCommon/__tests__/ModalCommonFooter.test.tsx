import { render } from '@testing-library/react';
import { emptyFunction } from 'shared/testsUtils';
import ModalCommonFooter from '../ModalCommonFooter';

const defaultProps = {
  onSubmit: emptyFunction,
  onCancel: emptyFunction,
};

describe('<ModalCommonFooter/>', () => {
  it('Should render ModalCommonFooter', () => {
    const { asFragment } = render(<ModalCommonFooter {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
