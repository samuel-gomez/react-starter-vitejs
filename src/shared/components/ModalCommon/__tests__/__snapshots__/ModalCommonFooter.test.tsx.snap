// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ModalCommonFooter/> > Should render ModalCommonFooter 1`] = `
<DocumentFragment>
  <div
    class="af-modal__footer"
  >
    <button
      class="btn af-btn af-btn--reverse"
      type="button"
    >
      <span
        class="af-btn__text"
      >
        Annuler
      </span>
    </button>
    <button
      class="btn af-btn af-btn--hasiconRight af-btn--success"
      type="button"
    >
      <span
        class="af-btn__text"
      >
        Confirmer
      </span>
      <i
        class="glyphicon glyphicon-arrowthin-right"
      />
    </button>
  </div>
</DocumentFragment>
`;
